packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,67 +165,213 @@ function makeSwingStore(dirPath, forceReset = false) {
+              itemCount -= 1;
+              assert(itemCount >= 0);
+              const result = line.toString();
+              if (skipCount > 0) {
+                skipCount -= 1;
+              } else {
+                yield result;
+              }
+            } else {
+              break;
+            }
           }
+        } finally {
+          const statusEnd = streamStatus.get(streamName);
+          assert(
+            statusEnd === readStatus,
+            X`can't read stream ${q(streamName)}, it's been closed`,
+          );
+          closeStream(streamName);
+          assert(itemCount === 0, X`leftover item count ${q(itemCount)}`);
         }
-        fs.closeSync(fd);
-      }
-    } catch (e) {
-      if (e.code !== 'ENOENT') {
-        throw e;
       }
+      return reader();
     }
   }
 
   /**
-   * Append an item to a stream.
+   * Obtain a writer for a stream.
    *
-   * @param {string} streamName  The stream to append to
-   * @param {string} item  The item to append to it
+   * @param {string} streamName  The stream to be written
+   *
+   * @returns {StreamWriter} a writer for the named stream
    */
-  function appendItem(streamName, item) {
-    assert.typeof(
-      streamName,
-      'string',
-      X`non-string stream name ${streamName}`,
+  function openWriteStream(streamName) {
+    insistStreamName(streamName);
+    const status = streamStatus.get(streamName);
+    assert(
+      status === 'unused' || !status,
+      X`can't write stream ${q(streamName)} because it's already in use`,
     );
-    assert(streamName.match(/^[-\w]+$/));
-    let fd = streamFds.get(streamName);
-    if (!fd) {
-      fd = fs.openSync(`${dirPath}/${streamName}`, 'a');
-      streamFds.set(streamName, fd);
+    streamStatus.set(streamName, 'write');
+
+    // XXX fdTemp is a workaround for a flaw in TypeScript's type inference
+    // It should be fd, which it should be changed to when and if they fix tsc.
+    let fdTemp = streamFds.get(streamName);
+    if (!fdTemp) {
+      const filePath = `${dirPath}/${streamName}`;
+      const mode = fs.existsSync(filePath) ? 'r+' : 'w';
+      fdTemp = fs.openSync(filePath, mode);
+      streamFds.set(streamName, fdTemp);
     }
+    const fd = fdTemp;
     activeStreamFds.add(fd);
-    fs.writeSync(fd, item);
-    fs.writeSync(fd, '\n');
+
+    /**
+     * Write to a stream.
+     *
+     * @param {string} item  The item to write
+     * @param {Object|null} position  The position to write the item
+     *
+     * @returns {Object} the new position after writing
+     */
+    function write(item, position) {
+      assert.typeof(item, 'string');
+      assert(
+        streamFds.get(streamName) === fd,
+        X`can't write to closed stream ${q(streamName)}`,
+      );
+      if (!position) {
+        position = STREAM_START;
+      }
+      const buf = encoder.encode(`${item}\n`);
+      fs.writeSync(fd, buf, 0, buf.length, position.offset);
+      fs.fsyncSync(fd);
```

Comment:
I don't think we need to fsync this now (and in general, letting the kernel decide when to flush the buffers is best for performance). Instead, I think we can wait for `commit()`. The rule is that any hard crash (including abrupt power loss or OS kernel crash) after the DB commit should include everything that was committed to, including the linear transcript files. So we need to push things to disk as hard as we can, at *some point* before the DB commit, but we don't need to do it any earlier than that.

So I'm thinking that swingstore `commit()` does a `activeStreamFds.forEach(fs.fsyncSync)`, and maybe `close()` on an individual stream does too (since we drop the fd and lose a handle to do it later), but `write()` does not.

(on rotating media, `fsync` is gonna take 1/5400RPM = 11ms, which would limit our crank rate to 90/s before we even get to spend any CPU time)

Reply:
Yeah, I agree.  I was thinking suspenders-and-belt, but it makes sense for `commit` to sync and `close` to close.